File: lecture/33-compilation/java/Main.java
public class Main {

    public static void main(String[] args) {
        System.out.println("Linked Lists!");
        
        Node list = Node.cons(1, Node.cons(2, Node.cons(3, null)));
        print_list(list);

        System.out.println("sum: " + sum(list));
    }

    public static void print_list(Node cursor) {
        while (cursor != null) {
            System.out.print(Node.first(cursor) + " -> ");
            cursor = Node.rest(cursor);
        }
        System.out.println("null");
    }

    public static int sum(Node cursor) {
        int sum = 0;
        while (cursor != null) {
            sum += Node.first(cursor);
            cursor = Node.rest(cursor);
        }
        return sum;
    }

}

File: lecture/33-compilation/c/main.c
#include <stdio.h>
#include <stdlib.h>

#include "node.h"

void print_list(Node* cursor);
int sum(Node* cursor);

int main() {
    printf("Linked Lists!\n");

    Node* list = cons(1, cons(2, cons(3, cons(4, NULL))));
    print_list(list);

    printf("Sum: %d\n", sum(list));

    return EXIT_SUCCESS;
}

void print_list(Node* cursor) {
    while (cursor != NULL) {
        printf("%d -> ", first(cursor));
        cursor = rest(cursor);
    }
    printf("NULL");
    printf("\n");
}

int sum(Node* cursor) {
    int sum = 0;
    while (cursor != NULL) {
        sum += first(cursor);
        cursor = rest(cursor);
    }
    return sum;
}
